File: src/main/scala/com/aoc2023/Day01.scala
package com.aoc2023

import scala.util.Using
import scala.io.Source
import scala.util.matching.Regex

object Day01 {

  val DIGITS = Seq(
    "zero",
    "one",
    "two",
    "three",
    "four",
    "five",
    "six",
    "seven",
    "eight",
    "nine"
  ).zipWithIndex.tail

  val numberPattern: Regex = "\\d".r

  def find_digits(str: String): Seq[String] = {
    numberPattern.findAllMatchIn(str).map(m => m.matched).toSeq
  }

  def find_digits_with_literals(str: String): Seq[String] = {
    str.zipWithIndex.flatMap { case (ch, idx) =>
      numberPattern.findFirstMatchIn(ch.toString) match {
        case Some(m) => Some(m.matched)
        case _ =>
          val s = str.splitAt(idx)._2
          DIGITS
            .find { case (num_str, num) => s.startsWith(num_str) }
            .map { case (num_str, num) => num.toString }
      }
    }
  }

  def get_calibration_number(digits: Seq[String]): Int = {
    if (digits.size > 0) {
      { digits.head + digits.last }.toInt
    } else {
      0
    }
  }

  def part1(path: String): Int = {
    Using.resource(Source.fromResource(path))(
      _.getLines()
        .map { line =>
          val digits = find_digits(line)
          get_calibration_number(digits)
        }
        .sum
    )
  }

  def part2(path: String): Int = {
    Using.resource(Source.fromResource(path))(
      _.getLines()
        .map { line =>
          val digits = find_digits_with_literals(line)
          get_calibration_number(digits)
        }
        .sum
    )
  }

  def main(args: Array[String]): Unit = {
    val path: String = "2023/day01.txt"
    println(s"Part 1: ${part1(path)}")
    println(s"Part 2: ${part2(path)}")
  }

}

File: src/python/day01/day01.py
import sys
import re

NUMBERS = {'one': '1', 'two': '2', 'three': '3', 'four': '4', 'five': '5',
           'six': '6', 'seven': '7', 'eight': '8', 'nine': '9'}


def create_calibration_value(digits):
    if len(digits) > 0:
        return int(digits[0] + digits[-1])
    else:
        return 0


def find_digits(line):
    return re.findall(r'\d', line)


def find_digits_with_literals(line):
    digits = []
    for i in range(len(line)):
        if line[i].isnumeric():
            digits.append(line[i])
        else:
            for num in NUMBERS.keys():
                if num == line[i: i + len(num)]:
                    digits.append(NUMBERS[num])
    return digits


def part_1(path):
    nums = []
    with open(path, 'r') as f:
        for line in f.readlines():
            digits = find_digits(line)
            num = create_calibration_value(digits)
            nums.append(num)
    return sum(nums)


def part_2(path):
    nums = []
    with open(path, 'r') as f:
        for line in f.readlines():
            digits = find_digits_with_literals(line)
            nums.append(create_calibration_value(digits))

    return sum(nums)


if __name__ == "__main__":
    path = sys.argv[1] if len(
        sys.argv) > 1 else "src/main/resources/2023/day01.txt"
    print(f"Part 1: {part_1(path)}")
    print(f"Part 2: {part_2(path)}")
